src/Text/Pandoc/Readers/Man.hs

```diff
@@ -0,0 +1,552 @@
+
+   Maintainer  : Yan Pashkovsky <yanp.bugz@gmail.com>
+   Stability   : WIP
+   Portability : portable
+
+Conversion of man to 'Pandoc' document.
+-}
+module Text.Pandoc.Readers.Man (readMan) where --testFile
+
+import Prelude
+import Control.Monad (liftM)
+import Control.Monad.Except (throwError)
+import Data.Char (isDigit, isUpper, isLower)
+import Data.Default (Default)
+import Data.Map (insert)
+import Data.Maybe (catMaybes, fromMaybe, isNothing)
+import Data.List (intersperse, intercalate)
+import qualified Data.Text as T
+
+import Text.Pandoc.Class (PandocMonad(..))
+import Text.Pandoc.Definition
+import Text.Pandoc.Error (PandocError (PandocParsecError))
+import Text.Pandoc.Logging (LogMessage(..))
+import Text.Pandoc.Options
+import Text.Pandoc.Parsing
+import Text.Pandoc.Shared (crFilter)
+import Text.Parsec hiding (tokenPrim)
+import Text.Parsec.Char ()
+import Text.Parsec.Pos (updatePosString)
+
+--
+-- Data Types
+--
+
+data FontKind = Regular | Italic | Bold | ItalicBold deriving Show
+
+data MacroKind = KTitle
+               | KCodeBlStart
+               | KCodeBlEnd
+               | KTab
+               | KTabEnd
+               | KSubTab
+               deriving (Show, Eq)
+
+type RoffStr = (String, FontKind)
+
+data ManToken = MStr RoffStr
+              | MLine [RoffStr]
+              | MMaybeLink String
+              | MEmptyLine
+              | MHeader Int [RoffStr]
+              | MMacro MacroKind [RoffStr]
+              | MUnknownMacro String [RoffStr]
+              | MComment String
+              deriving Show
+
+data EscapeThing = EFont FontKind
+                 | EChar Char
+                 | ENothing
+                 deriving Show
+
+data RoffState = RoffState { fontKind :: FontKind
+                           } deriving Show
+
+instance Default RoffState where
+  def = RoffState {fontKind = Regular}
+
+type ManLexer m = ParserT [Char] RoffState m
+type ManParser m = ParserT [ManToken] ParserState m
+
+---- debug functions
+{-
+import Text.Pandoc.Class (runIOorExplode)
+
+printPandoc :: Pandoc -> [Char]
+printPandoc (Pandoc m content) =
+  let ttl = "Pandoc: " ++ (show $ unMeta m)
+      cnt = intercalate "\n" $ map show content
+  in ttl ++ "\n" ++ cnt
+
+testStr :: String -> IO ()
+testStr str = do
+  pand <- runIOorExplode $ readMan def (T.pack str)
+  putStrLn $ printPandoc pand
+
+  
+testFile :: FilePath -> IO ()
+testFile fname = do
+  cont <- readFile fname
+  testStr cont
+-}
+----
+
+
+-- | Read man (troff) from an input string and return a Pandoc document.
+readMan :: PandocMonad m => ReaderOptions -> T.Text -> m Pandoc
+readMan opts txt = do
+  eithertokens <- readWithM lexMan def (T.unpack $ crFilter txt)
+  case eithertokens of
+    Right tokenz -> do
+      let state = def {stateOptions = opts} :: ParserState
+      eitherdoc <- readWithMTokens parseMan state tokenz
+      case eitherdoc of
+        Right doc -> return doc
+        Left e -> throwError e
+    Left e       -> throwError e
```

Comment:
for simple Either matches like l. 129, you might consider using `either`:
```haskell
either throwError return eitherdoc
```
You could even do:
```haskell
either throwError return <$> readWithMTokens parseMan state tokenz
```
and avoid the named `eitherdoc`.